src/viadot/sources/sftp.py

```diff
@@ -0,0 +1,296 @@
+            transport.connect(None, self.username, self.password)
+
+            self.conn = paramiko.SFTPClient.from_transport(transport)
+
+        else:
+            keyfile = StringIO(self.rsa_key)
+            mykey = paramiko.RSAKey.from_private_key(keyfile)
+            ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+            ssh.connect(self.hostname, username=self.username, pkey=mykey)
+            time.sleep(1)
+            self.conn = ssh.open_sftp()
+
+        self.logger.info("Connected to the SFTP server.")
+
+    @add_viadot_metadata_columns
+    def to_df(
+        self,
+        if_empty: str = "warn",
+        file_name: str | None = None,
+        sep: str = "\t",
+        columns: list[str] | None = None,
+    ) -> pd.DataFrame:
+        r"""Copy a remote file from the SFTP server and write it to Pandas dataframe.
+
+        Args:
+            if_empty (str, optional): What to do if a fetch produce no data.
+                Defaults to "warn
+            file_name (str, optional): File name to download.
+            sep (str, optional): The delimiter for the source file. Defaults to "\t".
+            columns (list[str], optional): List of columns to select from file.
+                Defaults to None.
+
+        Returns:
+            pd.DataFrame: The response data as a Pandas Data Frame plus viadot metadata.
+        """
+        byte_file = self._getfo_file(file_name=file_name)
+        byte_file.seek(0)
+
+        self._close_conn()
+
+        if Path(file_name).suffix == ".csv":
+            df = pd.read_csv(byte_file, sep=sep, usecols=columns)
+
+        elif Path(file_name).suffix == ".parquet":
+            df = pd.read_parquet(byte_file, usecols=columns)
+
+        elif Path(file_name).suffix == ".tsv":
+            df = pd.read_csv(byte_file, sep=sep, usecols=columns)
+
+        elif Path(file_name).suffix in [".xls", ".xlsx", ".xlsm"]:
+            df = pd.read_excel(byte_file, usecols=columns)
+
+        elif Path(file_name).suffix == ".json":
+            df = pd.read_json(byte_file)
+
+        elif Path(file_name).suffix == ".pkl":
+            df = pd.read_pickle(byte_file)
+
+        elif Path(file_name).suffix == ".sql":
+            df = pd.read_sql(byte_file)
+
+        elif Path(file_name).suffix == ".hdf":
+            df = pd.read_hdf(byte_file)
+
+        else:
+            message = (
+                f"Not able to read the file {Path(file_name).name}, "
+                + f"unsupported filetype: {Path(file_name).suffix}"
+            )
+            raise ValueError(message)
+
+        if df.empty:
+            self._handle_if_empty(
+                if_empty=if_empty,
+                message="The response does not contain any data.",
+            )
+        else:
+            self.logger.info("Successfully downloaded data from the SFTP server.")
+
+        return df
+
+    def _list_directory(self, path: str | None = None) -> list[str]:
```

Comment:
Why not just the standard `ls`? Also, this should take `recursive` as param rather than having a separate method just to implement the param.

```suggestion
    def _ls(self, path: str | None = None, recursive: bool = False) -> list[str]:
```

Reply:
This is a backlog task for refactoring the code.

Reply:
Changed only function name for now. https://github.com/dyvenia/viadot/pull/1039/commits/fd1712a6772061e8d52e5ec446791da8be84ba60

Reply:
Modified https://github.com/dyvenia/viadot/pull/1039/commits/770140a080a2e1604513ad917de69d5724f0eb82